src/QsCompiler/CompilationManager/CompilationUnit.cs

```diff
@@ -686,5 +701,71 @@ internal LocalDeclarations TryGetLocalDeclarations(FileContentManager file, Posi
+
+        /// <summary>
+        /// Returns a short tag name for the given source file. The tag is not unique. If multiple source files are
+        /// being tagged, make sure there are no name conflicts.
+        /// </summary>
+        /// <param name="source">The source file to get the tag for.</param>
+        /// <returns>A short tag name for the given source file.</returns>
+        internal static string GetTagForSource(string source) =>
+            Regex.Replace(Path.GetFileNameWithoutExtension(source), "[^A-Za-z0-9_]", "");
+
+        /// <summary>
+        /// Embeds the tag name into the qualified name.
+        /// </summary>
+        /// <param name="tag">The tag name to embed.</param>
+        /// <param name="name">The qualified name in which to embed the tag name.</param>
+        /// <returns>A new qualified name with the tag embedded into it.</returns>
+        internal static QsQualifiedName EmbedTag(string tag, QsQualifiedName name) =>
+            // TODO: We might need to change the format to make it easier to reverse later.
+            new QsQualifiedName(name.Namespace, NonNullable<string>.New($"__{name.Name.Value}_{tag}__"));
+
+        /// <summary>
+        /// Tags the names of imported internal callables and types with a unique identifier based on the path to their
+        /// assembly, so that they do not conflict with callables and types defined locally. Renames all references to
+        /// the tagged declarations found in the given callables and types.
+        /// </summary>
+        /// <param name="callables">The callables to rename and update references in.</param>
+        /// <param name="types">The types to rename and update references in.</param>
+        /// <returns>The tagged callables and types with references renamed everywhere.</returns>
+        private static (IEnumerable<QsCallable>, IEnumerable<QsCustomType>)
+            TagImportedInternalNames(IEnumerable<QsCallable> callables, IEnumerable<QsCustomType> types)
+        {
+            // Create a mapping from source file names to a short, unique identifying tag name.
+            var tags =
+                callables.Select(callable => callable.SourceFile.Value)
+                .Concat(types.Select(type => type.SourceFile.Value))
+                .Distinct()
+                .GroupBy(GetTagForSource)
+                .SelectMany(group =>
+                    group.Count() == 1
+                    ? new[] { (key: group.Single(), value: group.Key) }
+                    : group.Select((source, index) => (key: source, value: group.Key + index)))
+                .ToImmutableDictionary(item => item.key, item => item.value);
+
+            ImmutableDictionary<QsQualifiedName, QsQualifiedName> GetMappingForSourceGroup(
+                IGrouping<string, (QsQualifiedName name, string source, AccessModifier access)> group) =>
+                // TODO: Is there another way besides file extension to check if the source file is a reference?
```

Comment:
Not currently. What I meant to do, however, is to generated a Q# attribute for declarations that come from references to tag them. I'll make a task under clean-up to do that.

Reply:
It looks like I can use `Externals.Declarations.ContainsKey(source)` instead.